### PR TITLE
Fix incorrect `into_buffers` for UnionArray

### LINKED ISSUE #1547: Output of `filter` on UnionArrays fails validation
**Describe the bug**
The `force_validate` check added in https://github.com/apache/arrow-rs/pull/1546 revealed that the `UnionArrays` produced from the `filter` kernel fails validation.

It is not clear if the validation is incorrect or the filter kernel is incorrect

**To Reproduce**
Uncomment the
```rust
    #[cfg(not(feature = "force_validate"))]
``` 

And then run the test like:
```shell
cargo test --features=force_validate -p arrow
```

lines for the following tests and the will fail:

```
---- compute::kernels::filter::tests::test_filter_union_array_sparse stdout ----
thread 'compute::kernels::filter::tests::test_filter_union_array_sparse' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidArgumentError("Expected 1 buffers in array of type Union([Field { name: \"A\", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"B\", data_type: Float64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: None }], Sparse), got 2")', arrow/src/array/data.rs:301:34

---- compute::kernels::filter::tests::test_filter_union_array_sparse_with_nulls stdout ----
thread 'compute::kernels::filter::tests::test_filter_union_array_sparse_with_nulls' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidArgumentError("Expected 1 buffers in array of type Union([Field { name: \"A\", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: None }, Field { name: \"B\", data_type: Float64, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: None }], Sparse), got 2")', arrow/src/array/data.rs:301:34
```

**Expected behavior**
The tests should pass when force_validate is on

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #1547.

# Rationale for this change
 
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by tustvold — APPROVED** [review]

**Review by tustvold — COMMENTED** [review]
I appreciate https://github.com/apache/arrow-rs/pull/1546 will add some test coverage, but I think a simple explicit test couldn't hurt

**Comment by viirya** [comment]
Thanks @tustvold . Okay, I will add a test for this.

File: arrow/src/array/data.rs [hunk 1]
```diff
@@ -2594,4 +2600,21 @@ mod tests {
 
         assert_eq!(&struct_array_slice, &cloned);
     }
+
+    #[test]
+    fn test_into_buffers() {
+        let data_types = vec![
+            DataType::Union(vec![], UnionMode::Sparse),
+            DataType::Union(vec![], UnionMode::Sparse),
+        ];
```
Comment (viirya) [inline arrow/src/array/data.rs:2609]: We can add more datatypes if needed.

File: arrow/src/compute/kernels/filter.rs [hunk 2]
```diff
@@ -1692,9 +1692,6 @@ mod tests {
     }
 
     #[test]
-    // Fails when validation enabled
-    // https://github.com/apache/arrow-rs/issues/1547
-    #[cfg(not(feature = "force_validate"))]
     fn test_filter_union_array_sparse() {
```
Comment (viirya) [inline arrow/src/compute/kernels/filter.rs:1695]: Re-enabled them here. cc @alamb

**Review by tustvold — APPROVED** [review]
Event better now 😃

**Comment by viirya** [comment]
Thank you @tustvold !

File: arrow/src/compute/kernels/filter.rs (hunk 2 above)
Comment (alamb) [reply arrow/src/compute/kernels/filter.rs:1695]: ❤️ so nice -- thank you @viirya

**Pull request merged**